### PR TITLE
Improve documentation on security topics

### PR DESCRIPTION
**Pull request by desislavaa**
**Description**
Documentation improvements on the follwing aspects:
- Consumer-provider flow documentaton (SaaS app)
- Consumer-provider flow documentaton (Integration System)
- Internal Authentication
- Filters

### CODE REVIEW

File: chart/compass/templates/filters/limit-request-payload-size.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+apiVersion: networking.istio.io/v1alpha3
```
Comment (desislavaa) [inline chart/compass/templates/filters/limit-request-payload-size.yaml:1]: remove

File: docs/compass/03-01-security.md [hunk 2]
```diff
@@ -246,3 +290,58 @@ Scopes are added to the authentication session in Tenant Mapping Handler. The ha
...
+
+**Scopes**
+Since there will be two context providers, there will also be two pair of scopes sets - one from the matched external certificate context provider, and one from the access level context provider.
+In this case, the scopes from the external certificate context provider (for consumer type Integration System) will be taken into account.
+
+##### Runtimes
+As Integration Systems are usually too powerful, and we should be careful who gets Integration System access, a more flexible and secured way of achieving the same scenarios was introduced.
+First, the multi-tenant applications should be registered as a "special" runtime in Compass, which contains a special label, which Compass uses for distinguishing multi-tenant and ordinary runtimes.
+
+When the multi-tenant system is represented as a Runtime, its tenant access is managed from an outside service, which communicates with the Tenant Fetcher deployment, and grants the consumer tenant (the tenant where the multi-tenant runtime is registered) access to resources in the provider tenant, which will be of type `subaccount`.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:318]: consumer and provider are switched here.

File: docs/compass/03-01-security.md [hunk 3]
```diff
@@ -246,3 +290,58 @@ Scopes are added to the authentication session in Tenant Mapping Handler. The ha
...
+The Integration Systems can use OAuth credentials for authentication (created via the `requestClientCredentialsForIntegrationSystem` mutation), or client certificate, issued by an external issuer, which is trusted by Compass.
+
+The second authentication mechanism is plugged into Compass Connector - it can be configured to trust a specific certificate subject, and based on that subject, to set the Consumer type to _Integration System_, and grant it access to a set of tenant types - it is a bit more restrictive than the one that uses OAuth client.
+The provider tenant is specified in the `Tenant` header.
+
+**Compass Director Flow:**
+
+The flow is basically the one described above for externally issued certificate. The only difference is that the Tenant Mapping handler will match two object context providers, hence the tenants in this request will be 4 in total - internal and external provider tenants, and internal and external consumer tenants.
+If the consumer tenant has access to the type of tenant the provider belongs to, then the request is authorized successfully.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:308]: consumer and provider are switched here.

File: docs/compass/03-01-security.md [hunk 4]
```diff
@@ -246,3 +290,58 @@ Scopes are added to the authentication session in Tenant Mapping Handler. The ha
...
+
+When the multi-tenant system is represented as a Runtime, its tenant access is managed from an outside service, which communicates with the Tenant Fetcher deployment, and grants the consumer tenant (the tenant where the multi-tenant runtime is registered) access to resources in the provider tenant, which will be of type `subaccount`.
+
+Once the trust between the tenants is established, the user can access any resource which resides in in its tenant from the external multi-tenant system.
+
+**Compass Director Flow:**
+
+_Not productive yet_
+
+**Compass ORD Service Flow:**
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:326]: This one is productive already.

File: docs/compass/03-01-security.md [hunk 5]
```diff
@@ -246,3 +290,58 @@ Scopes are added to the authentication session in Tenant Mapping Handler. The ha
...
+This flow is also almost the same as the externally issued certificate one.
+Just like above, two context providers are matched - the externally issued certificate one, and a custom authentication one with JWT.
+Compass will validate that the consumer tenant has access to the provider tenant.
+
+**Scopes**
+The first context provider will return Runtime scopes, and the second one will return the scopes present in the JWT. We should use the ones from the token.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:332]: Not sure this is true, let's sync offline.

File: docs/compass/03-02-filters.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+Istio also provides a tracing mechanism, which is described in details [here](https://istio.io/latest/about/faq/distributed-tracing/#how-to-support-tracing). The most important thing is that it uses `x-request-id` to do so, hence `x-request-id` is the correlation header we will use for requests across our components.
+
+In case an external service has attached a correlation ID header, we will get its value, and set it to `x-request-id` - that's what the `compass-correlation-headers-rewrite` Envoy filter does.
+It is required because there is no strict convention regarding the correlation ID header name. The default ones we're looking for are listed [here](https://github.com/kyma-incubator/compass/blob/d429d01b34eb1a7512c9613dff2bc27d6c814857/chart/compass/values.yaml#L304).
+
+### Disable Istio Retries Filter
+This filter sets the `x-envoy-max-retries` header to 0, which will disable the default retry behavior, because it is not flexible enough with the currently supported Istio version. 
+
+### ORY Oathkeeper Timeout Filter
+This filter was introduced because there are no timeouts on ARY Oathkeeper side, but we want to cancel requests that take more time than usual.
```
Comment (DimitarPetrov) [inline docs/compass/03-02-filters.md:16]: typo ARY -> ORY

File: docs/compass/03-01-security.md [hunk 7]
```diff
@@ -246,3 +290,58 @@ Scopes are added to the authentication session in Tenant Mapping Handler. The ha
...
+
+Compass has a separate security layer to take care of internal component communication (e.g. from ORY Oathkeeper to Director). 
+
+It is based on Istio `RequestAuthentications` and `AuthorizationPolicies` that require service account token from the caller as an additional X-Authorization header. The service account in question is the one bound to the source workload.
+The `RequestAuthentication` verifies that the token is correctly signed by the Kubernetes JWKS. 
+The `AuthorizationPolicy` makes sure that routes such as `/healthz` are accessible without any authentication, and all other routes require properly signed tokens.
+
+Workloads that previously required some form of authentication, e.g. ORY Oathkeeper tokens for Director, or one-time tokens for Connector, still require them as a separate header.
+
+![](./assets/internal-auth.png)
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:344]: Not rendered.

**Review by DimitarPetrov — CHANGES_REQUESTED** [review]

**Review by nkanchev — APPROVED** [review]

File: docs/compass/03-01-security.md [hunk 8]
```diff
@@ -1,12 +1,18 @@
...
  - OAuth 2.0
  - Client certificates (mTLS)
- - JWT token issued by identity service
+    - Issued by the Compass Connector
+    - Issued by external certificate authority
+ - JSON Web Token (JWT) issued by identity service
  - One-time token
+ 
+Additionally, there are **two consumer-provider** flows, in which an external multi-tenant system wants to manage resources (for example, bundle instance auths) on behalf of a tenant in Compass. These are the following:
+ - SaaS applications, which are modeled like runtimes, can access Compass resources with an externally-issued certificate, and a JWT where the provider tenant is present.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:12]: ... and a JWT where the `consumer` tenant is present

File: docs/compass/03-01-security.md [hunk 9]
```diff
@@ -16,30 +22,46 @@ The following diagram represents the architecture of the security in Compass:
...
+1. The tenant-mapping handler extracts them from the token's claims and verifies that the tenant exists.
+
+#### Consumer-Provider Flows
+As mentioned above, there are use cases where you can manage Compass resources from a different multi-tenant system, which shares the same tenancy model and knows the external tenant IDs, which are also used by Compass. In this case, Compass trusts those systems and allows them to manage resources on behalf of the user. The following options are valid for this flow:
+- The multi-tenant system is represented as an Integration System. In this case, the provider tenant is specified in the `Tenant` header. The scopes which are granted to the request, are the ones assigned to integration systems, by default. They are granted by the context provider, mentioned above.
+- The multi-tenant system is represented as a Runtime. In this case, the second context provider that matches is a custom authenticator, where the provider tenant is part of a JWT, along with the scopes that are granted. The scopes are taken from the context provider in this case.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:62]: ... where the `consumer` tenant is part of a JWT ...

File: docs/compass/03-01-security.md [hunk 10]
```diff
@@ -16,30 +22,46 @@ The following diagram represents the architecture of the security in Compass:
...
+
+Both cases feature two context providers, which are used in a pair. One of context providers must be an externally-issued certificate context provider. It extracts the consumer tenant ID from the certificate. Later on, Compass checks if the consumer tenant has access to the provider tenant. For more information, see the [Authentication Flows](03-01-security.md#authentication-flows) section in this document.
```
Comment (DimitarPetrov) [inline docs/compass/03-01-security.md:64]: In the whole doc consumer and provider seems to be switched based on compass perspective. 

provider - the tenant providing system metadata (like ORD provider)
consumer - the tenant consuming system metadata (like ORD consumer)

However, the most common use of those terms in the SaaS registry world are from the SaaS app's perspective:

provider - the SaaS application that provides some features to a client
consumer - the client that consumes those features

Maybe we can stick to this terminology as most widely used in other resources.

**Review by DimitarPetrov — CHANGES_REQUESTED** [review]

**Review by DimitarPetrov — APPROVED** [review]

**Review by nkanchev — APPROVED** [review]

**Pull request merged**